src/ion/data_structures.rs

```diff
@@ -13,19 +13,139 @@
 //! Data structures for backtracking allocator.
 
 use super::liveranges::SpillWeight;
+use super::moves::MoveCtx;
 use crate::cfg::CFGInfo;
 use crate::index::ContainerComparator;
 use crate::indexset::IndexSet;
 use crate::{
-    define_index, Allocation, Block, Edit, Function, FxHashSet, MachineEnv, Operand, PReg,
-    ProgPoint, RegClass, VReg,
+    define_index, Allocation, Block, Edit, Function, FxHashMap, FxHashSet, MachineEnv, Operand,
+    Output, PReg, ProgPoint, RegClass, VReg, VecExt,
 };
-use alloc::collections::BTreeMap;
+//use alloc::collections::BTreeMap;
+use alloc::collections::VecDeque;
+use alloc::rc::Rc;
 use alloc::string::String;
 use alloc::vec::Vec;
 use core::cmp::Ordering;
+use core::convert::identity;
 use core::fmt::Debug;
-use smallvec::{smallvec, SmallVec};
+use core::ops::{Deref, DerefMut};
+use smallvec::SmallVec;
+
+#[derive(Debug, Clone, Default)]
+pub struct Bump(Rc<bumpalo::Bump>);
+
+impl Bump {
+    pub(crate) fn get_mut(&mut self) -> Option<&mut bumpalo::Bump> {
+        Rc::get_mut(&mut self.0)
+    }
+}
+
+// simply delegating beause `Rc<bumpalo::Bump>` does not implement `Allocator`
+unsafe impl allocator_api2::alloc::Allocator for Bump {
+    fn allocate(
+        &self,
+        layout: core::alloc::Layout,
+    ) -> Result<core::ptr::NonNull<[u8]>, allocator_api2::alloc::AllocError> {
+        self.0.deref().allocate(layout)
+    }
+
+    unsafe fn deallocate(&self, ptr: core::ptr::NonNull<u8>, layout: core::alloc::Layout) {
+        self.0.deref().deallocate(ptr, layout);
+    }
+
+    fn allocate_zeroed(
+        &self,
+        layout: core::alloc::Layout,
+    ) -> Result<core::ptr::NonNull<[u8]>, allocator_api2::alloc::AllocError> {
+        self.0.deref().allocate_zeroed(layout)
+    }
+
+    unsafe fn grow(
+        &self,
+        ptr: core::ptr::NonNull<u8>,
+        old_layout: core::alloc::Layout,
+        new_layout: core::alloc::Layout,
+    ) -> Result<core::ptr::NonNull<[u8]>, allocator_api2::alloc::AllocError> {
+        self.0.deref().grow(ptr, old_layout, new_layout)
+    }
+
+    unsafe fn grow_zeroed(
+        &self,
+        ptr: core::ptr::NonNull<u8>,
+        old_layout: core::alloc::Layout,
+        new_layout: core::alloc::Layout,
+    ) -> Result<core::ptr::NonNull<[u8]>, allocator_api2::alloc::AllocError> {
+        self.0.deref().grow_zeroed(ptr, old_layout, new_layout)
+    }
+
+    unsafe fn shrink(
+        &self,
+        ptr: core::ptr::NonNull<u8>,
+        old_layout: core::alloc::Layout,
+        new_layout: core::alloc::Layout,
+    ) -> Result<core::ptr::NonNull<[u8]>, allocator_api2::alloc::AllocError> {
+        self.0.deref().shrink(ptr, old_layout, new_layout)
+    }
+}
+
+#[derive(Clone, Debug)]
+pub struct BTreeMap<K, V> {
+    pub values: Vec<(K, V)>,
```

Comment:
I'm fairly worried about worst-case behavior here; to the extent that I don't know if I want to accept this replacement of the BTreeMap unless we can make a clear argument why we will never hit quadratic behavior. We've had fuzzbugs in the past that uncover quadratic cases and we have to protect against them (we consider it a DoS vector).